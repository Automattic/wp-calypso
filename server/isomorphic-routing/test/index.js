/**
 * Internal dependencies
 */
import { getNormalizedPath } from '..';

describe( 'getNormalizedPath', () => {
	test( 'should return expected results for a variety of inputs', () => {
		const pathnames = [ '/', '/shortpath', '/a/longer/path', '/specialchars/%2F%3F%26%3D%25' ];
		const queries = [
			{},
			{ a: 'query' },
			{ '?tricky=query': '' },
			{ key: '=tricky&value=;)' },
			{ emptyValue: '' },
			{ nullValue: null },
		];

		// Cartesian product: paths * queries
		pathnames.forEach( pathname =>
			queries.forEach( query => expect( getNormalizedPath( pathname, query ) ).toMatchSnapshot() )
		);
	} );

	test( 'should return pathname for routes with no query', () => {
		const pathname = '/my/path';
		const query = {};

		expect( getNormalizedPath( pathname, query ) ).toBe( '/my/path' );
	} );

	test( 'should return normalized path for one query param', () => {
		const pathname = '/';
		const query = { cache_me: '1' };

		expect( getNormalizedPath( pathname, query ) ).toBe( '/?cache_me=1' );
	} );

	test( 'should return normalized pathname for multiple query params', () => {
		const pathname = '/';
		const query = { cache_me: '1', and_me: '2', me_too: '3' };

		expect( getNormalizedPath( pathname, query ) ).toBe( '/?and_me=2&cache_me=1&me_too=3' );
	} );

	test( 'should return a stable key pathname', () => {
		const pathname = '/';
		const query = { a: '1', b: '2' };
		const querySwapped = { b: '2', a: '1' };

		expect( getNormalizedPath( pathname, query ) ).toBe(
			getNormalizedPath( pathname, querySwapped )
		);
	} );

	test( 'should handle "empty" query params coherently', () => {
		expect( getNormalizedPath( '/', { empty: '' } ) ).toBe(
			getNormalizedPath( '/', { empty: null } )
		);
	} );

	test( 'should produce unique results', () => {
		expect( getNormalizedPath( '/', { key: 'val' } ) ).not.toBe(
			getNormalizedPath( '/', { 'key=val': '' } )
		);
	} );
} );
