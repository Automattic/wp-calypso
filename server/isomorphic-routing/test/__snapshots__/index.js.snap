// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNormalizedPath should return expected results for a variety of inputs 1`] = `""`;

exports[`getNormalizedPath should return expected results for a variety of inputs 2`] = `"?a=query"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 3`] = `"?%3Ftricky%3Dquery="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 4`] = `"?key=%3Dtricky%26value%3D%3B%29"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 5`] = `"?emptyValue="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 6`] = `"?nullValue="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 7`] = `"/"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 8`] = `"/?a=query"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 9`] = `"/?%3Ftricky%3Dquery="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 10`] = `"/?key=%3Dtricky%26value%3D%3B%29"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 11`] = `"/?emptyValue="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 12`] = `"/?nullValue="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 13`] = `"/shortpath"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 14`] = `"/shortpath?a=query"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 15`] = `"/shortpath?%3Ftricky%3Dquery="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 16`] = `"/shortpath?key=%3Dtricky%26value%3D%3B%29"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 17`] = `"/shortpath?emptyValue="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 18`] = `"/shortpath?nullValue="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 19`] = `"/a/longer/path"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 20`] = `"/a/longer/path?a=query"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 21`] = `"/a/longer/path?%3Ftricky%3Dquery="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 22`] = `"/a/longer/path?key=%3Dtricky%26value%3D%3B%29"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 23`] = `"/a/longer/path?emptyValue="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 24`] = `"/a/longer/path?nullValue="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 25`] = `"/specialchars/%2F%3F%26%3D%25"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 26`] = `"/specialchars/%2F%3F%26%3D%25?a=query"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 27`] = `"/specialchars/%2F%3F%26%3D%25?%3Ftricky%3Dquery="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 28`] = `"/specialchars/%2F%3F%26%3D%25?key=%3Dtricky%26value%3D%3B%29"`;

exports[`getNormalizedPath should return expected results for a variety of inputs 29`] = `"/specialchars/%2F%3F%26%3D%25?emptyValue="`;

exports[`getNormalizedPath should return expected results for a variety of inputs 30`] = `"/specialchars/%2F%3F%26%3D%25?nullValue="`;
